include/seqan3/io/detail/misc_input.hpp

```diff
@@ -73,32 +75,50 @@ template <char_concept char_t>
 inline auto make_secondary_istream(std::basic_istream<char_t> & primary_stream, std::filesystem::path & filename)
     -> std::unique_ptr<std::basic_istream<char_t>, std::function<void(std::basic_istream<char_t>*)>>
 {
+    assert(primary_stream.good());
+
     // don't assume ownership
     constexpr auto stream_deleter_noop     = [] (std::basic_istream<char_t> *) {};
     // assume ownership
     [[maybe_unused]] constexpr auto stream_deleter_default  = [] (std::basic_istream<char_t> * ptr) { delete ptr; };
 
     // extract "magic header"
     std::istreambuf_iterator<char_t> it{primary_stream};
-    constexpr size_t magic_header_size = 4;
-    std::array<char, magic_header_size> magic_number;
-    for (size_t i = 0; i < magic_header_size; ++i)
+    std::array<char, magic_header<bgzf_compression>.size()> magic_number{}; // Largest magic header from bgzf
+    size_t read_chars = 0;
+    for (; read_chars < magic_number.size(); ++read_chars)
     {
-        if (it == std::istreambuf_iterator<char_t>{}) // file is smaller than 4 bytes -> not compressed, could be legal
-            return {&primary_stream, stream_deleter_noop};
+        if (it == std::istreambuf_iterator<char_t>{})
+            break;
 
-        magic_number[i] = *it;
+        magic_number[read_chars] = *it;
         ++it;
     }
 
-    // restore original stream content
-    for (size_t i = 0; i < magic_header_size; ++i)
-        primary_stream.putback(magic_number[magic_header_size - 1 - i]);
+    // unget all read chars.
+    for (size_t i = 0 ; i < read_chars; ++i)
+        primary_stream.unget();
+
+    // file is smaller than 4 bytes -> not compressed, could be legal
+    if (read_chars < 4)
+        return {&primary_stream, stream_deleter_noop};
```

Comment:
Why 4 ? This was originally the header read, but now you read more. The problem is more complex however. A file might be smaller that the magic header of bgzf (18 bytes) and still be valid uncompressed  OR compressed. E.g. `> \n` compressed in zstd is 16 bytes.

Reply:
Maybe removing this check will just be ok :thinking:

Reply:
I think everything is still correct. We get at most `read_chars` chars. If this is less than 4 we assume an uncompressed file. If `read_char` is exactly 18 (number of bytes we read at most) and it is a bgzf header we read bgzf. Otherwise we check if it is gz, bz2, or zsd. In this case we have read at least 4 bytes (size of magic header of zstd) and `read_chars` <= 18 bytes. If we don't check for at least 4 bytes, then I need to check again if `starts_with` is correct. Then of course I can remove the check of 4 bytes and return the uncompressed primary stream at the end. I simply moved it out since before we were not ungetting the read chars from the stream if it was less than 4 bytes.